### PR TITLE
feat: allow scaffolder tasks pagination in the backend

### PR DESCRIPTION
**Pull request by drodil**
## Hey, I just made a Pull Request!

Still requires the UI to take use of this. Relates to #25856

#### :heavy_check_mark: Checklist

- [x] A changeset describing the change and affected packages. ([more info](https://github.com/backstage/backstage/blob/master/CONTRIBUTING.md#creating-changesets))
- [ ] Added or updated documentation
- [x] Tests for new functionality and regression tests for bug fixes
- [ ] Screenshots attached (for UI changes)
- [x] All your commits have a `Signed-off-by` line in the message. ([more info](https://github.com/backstage/backstage/blob/master/CONTRIBUTING.md#developer-certificate-of-origin))

### CODE REVIEW

**Comment by jhaals** [comment]
Thank you for looking at this @drodil! It's definitely not ideal to run without pagination.
I took a look at the api-report and this is currently a breaking change in the API meaning we would have to release 2.0 of scaffolder. Could you look at alternatives to implement this that does not break the existing API. I'm thinking some optional `pagination` object part of `list` or perhaps another list endpoint entierly if that could be made non-breaking.

**Comment by drodil** [comment]
Hm, I don't see why it's a breaking change as all these parameters to API are optional?

**Comment by jhaals** [comment]
> Hm, I don't see why it's a breaking change as all these parameters to API are optional?

Changing for `createdBy` from string to string[] does break the API even if supplying the field is optional.

**Comment by drodil** [comment]
Oh yeah true. Maybe it could be `string | string[]`?

**Comment by drodil** [comment]
@jhaals added back the option to filter with single `ownedBy` and `status`, WDYT?

File: plugins/scaffolder-node/api-report.md [hunk 1]
```diff
@@ -315,8 +315,18 @@ export interface TaskBroker {
...
+    status?: TaskStatus | TaskStatus[];
+    limit?: number;
+    offset?: number;
+    order?: {
+      order: 'asc' | 'desc';
+      field: string;
+    }[];
+  }): Promise<{
     tasks: SerializedTask[];
+    totalTasks: number;
```
Comment (jhaals) [inline plugins/scaffolder-node/api-report.md:329]: This is also a breaking change given that anyone already implementing list would have to start supplying this field.

**Review by jhaals — COMMENTED** [review]
I feel that this API change need some additional input from @benjdlambert and the SIG for how to best do this. It might be that being able to provide string or string[] is good and making totalTasks also optional to return but there might also be a case for a separate pagination object inside options or a separate function entirely. I'm not sure what's best tbh

**Comment by drodil** [comment]
Ok, maybe something to discuss in SIG. We are currently having around ~3000 template runs and the page is taking forever to load. At some point, it will cause failure due to too large JSON payload so I hope this (or whatever the decision is) will get merged soonish.

File: plugins/scaffolder-backend/api-report.md [hunk 2]
```diff
@@ -428,8 +428,18 @@ export class DatabaseTaskStore implements TaskStore {
   // (undocumented)
   heartbeatTask(taskId: string): Promise<void>;
   // (undocumented)
-  list(options: { createdBy?: string; status?: TaskStatus_2 }): Promise<{
```
Comment (benjdlambert) [inline plugins/scaffolder-backend/api-report.md:431]: I think it's best to leave these options as required, otherwise it's a breaking change to the API for implementers of this API. It's a mistake that it's required for sure, but let's not break it just yet.

File: plugins/scaffolder-backend/api-report.md [hunk 3]
```diff
@@ -428,8 +428,18 @@ export class DatabaseTaskStore implements TaskStore {
...
+  list(options?: {
+    createdBy?: string | string[];
+    status?: TaskStatus_2 | TaskStatus_2[];
+    limit?: number;
```
Comment (benjdlambert) [inline plugins/scaffolder-backend/api-report.md:434]: Maybe we should move `limit` and `offset` into a `pagination` option? Thoughts?

File: plugins/scaffolder-backend/api-report.md [hunk 4]
```diff
@@ -428,8 +428,18 @@ export class DatabaseTaskStore implements TaskStore {
...
+    offset?: number;
+    order?: {
+      order: 'asc' | 'desc';
+      field: string;
+    }[];
+  }): Promise<{
     tasks: SerializedTask_2[];
+    totalTasks: number;
```
Comment (benjdlambert) [inline plugins/scaffolder-backend/api-report.md:442]: I think we need to make this optional in the API too, otherwise again, it's breaking for implementers of the API.

**Review by benjdlambert — COMMENTED** [review]
@drodil I think this is heading the right direction, just a few changes to the API interface we need to make to make this a non-breaking change.

File: plugins/scaffolder-backend/api-report.md (hunk 3 above)
Comment (drodil) [reply plugins/scaffolder-backend/api-report.md:434]: No strong opinion either way; do you think there's going to be more pagination options?

File: plugins/scaffolder-backend/api-report.md (hunk 2 above)
Comment (drodil) [reply plugins/scaffolder-backend/api-report.md:431]: Ok, will fix

File: plugins/scaffolder-backend/src/scaffolder/tasks/StorageTaskBroker.ts [hunk 5]
```diff
@@ -279,18 +279,18 @@ export class StorageTaskBroker implements TaskBroker {
...
     if (!this.storage.list) {
       throw new Error(
         'TaskStore does not implement the list method. Please implement the list method to be able to list tasks',
       );
     }
-    return await this.storage.list({
-      createdBy: options?.createdBy,
-      status: options?.status,
-    });
+    return await this.storage.list(options ?? {});
```
Comment (jhaals) [inline plugins/scaffolder-backend/src/scaffolder/tasks/StorageTaskBroker.ts:293]: is `?? {}` required?

Comment (drodil) [reply plugins/scaffolder-backend/src/scaffolder/tasks/StorageTaskBroker.ts:293]: Yes because the options is required

File: plugins/scaffolder-backend/src/service/helpers.ts [hunk 6]
```diff
@@ -107,3 +107,23 @@ export async function findTemplate(options: {
...
   return template as TemplateEntityV1beta3;
 }
+
+/**
+ * Takes a single unknown parameter and makes sure that it's a single string or
+ * an array of strings, and returns as an array.
+ */
+export function parseStringsParam(
+  param: unknown,
+  ctx: string,
```
Comment (acierto) [inline plugins/scaffolder-backend/src/service/helpers.ts:117]: Maybe to name it as `parameterName` to make it easier to understand without looking up for the usage to get what is `ctx`?

File: plugins/scaffolder-backend/src/service/router.ts [hunk 7]
```diff
@@ -569,41 +574,71 @@ export async function createRouter(
...
+          const ret = Number.parseInt(val, 10);
+          if (isNaN(ret)) {
+            throw new InputError(
+              `Invalid limit parameter "${val}", expected a number`,
+            );
+          }
+          return ret;
+        });
+
+        const offset = parseStringsParam(req.query.offset, 'offset')?.map(
```
Comment (acierto) [inline plugins/scaffolder-backend/src/service/router.ts:618]: It does look like a copy paste of `limit`, can we extract it to some internal function?

File: plugins/scaffolder-backend/src/service/router.ts [hunk 8]
```diff
@@ -569,41 +574,71 @@ export async function createRouter(
...
+          createdBy,
+          status: status ? (status as TaskStatus[]) : undefined,
+          order,
+          limit: limit ? limit[0] : undefined,
+          offset: offset ? offset[0] : undefined,
+        });
+
+        res.status(200).json(tasks);
+      } catch (e) {
+        console.error(e);
```
Comment (acierto) [inline plugins/scaffolder-backend/src/service/router.ts:640]: I suppose you wanted to do here something else 😅

**Review by acierto — CHANGES_REQUESTED** [review]

File: plugins/scaffolder-backend/src/service/helpers.ts (hunk 6 above)
Comment (drodil) [reply plugins/scaffolder-backend/src/service/helpers.ts:117]: This is a nice copy-paste from the catalog, but yes, it's not that clear

File: plugins/scaffolder-backend/src/service/router.ts (hunk 8 above)
Comment (drodil) [reply plugins/scaffolder-backend/src/service/router.ts:640]: True, the whole try is unnecessary

**Review by jhaals — CHANGES_REQUESTED** [review]
I had another round of review with Ben. There are still changes that are considered breaking in the api-reports. Changing `string` to `string | string[]`, means that implementers would need to also handle the array case which they would not without changes. Same things goes for status now also accepting an array.
We could instead deprecate the old fields then create a new optional `filter` option accepting arrays.

**Comment by drodil** [comment]
@jhaals ok, I can add the `filters` object to the interface though it unnecessarily duplicates things.

**Comment by jhaals** [comment]
> @jhaals ok, I can add the `filters` object to the interface though it unnecessarily duplicates things.

I don't know of a better way than to deprecate the old and move to a filters object, happy to hear other suggestions :)

**Comment by drodil** [comment]
Fine for me at this point, maybe the breaking change will come with scaffolder 2.0 then? :)

**Comment by jhaals** [comment]
> Fine for me at this point, maybe the breaking change will come with scaffolder 2.0 then? :)

For sure! The fields you `@deprecate` now should definitely be dropped in 2.0 :)

**Comment by drodil** [comment]
@jhaals great, I will add the deprecated tags and it's ready for another round!

File: plugins/scaffolder-node/src/tasks/types.ts [hunk 9]
```diff
@@ -181,7 +181,20 @@ export interface TaskBroker {
   get(taskId: string): Promise<SerializedTask>;
 
   list?(options?: {
+    /**
+     * @deprecated Please use the `filters` parameter instead
```
Comment (jhaals) [inline plugins/scaffolder-node/src/tasks/types.ts:185]: Realized that this does not show up in the API report and editor as strikethrough unfortunately.

The thing we can do is to overload list and deprecate the old list. The new list method need to be added before the deprecated one to show up.

```ts
  list?(options?: {
    filters?: {
      createdBy?: string | string[];
      status?: TaskStatus | TaskStatus[];
    };
    limit?: number;
    offset?: number;
    order?: { order: 'asc' | 'desc'; field: string }[];
  }): Promise<{ tasks: SerializedTask[]; totalTasks?: number }>;

  /**
   * @deprecated Make sure to pass `createdBy` and `status` as filters to the `list` method instead
   */
  list?(options?: {
    createdBy?: string;
    status?: TaskStatus;
    filters?: {
      createdBy?: string | string[];
      status?: TaskStatus | TaskStatus[];
    };
    limit?: number;
    offset?: number;
    order?: { order: 'asc' | 'desc'; field: string }[];
  }): Promise<{ tasks: SerializedTask[]; totalTasks?: number }>;
```

File: .changeset/early-drinks-kneel.md [hunk 10]
```diff
@@ -0,0 +1,6 @@
+---
+'@backstage/plugin-scaffolder-backend': minor
+'@backstage/plugin-scaffolder-node': minor
+---
+
+Allow paginating scaffolder tasks in the backend
```
Comment (jhaals) [inline .changeset/early-drinks-kneel.md:6]:
```suggestion
Added pagination support for listing of tasks and the ability to filter on several users and task statuses.
```
Feel free to tweak this further :)

**Review by jhaals — CHANGES_REQUESTED** [review]
Nice work! Found some minor things with deprecations and there are some tests failing currently :)

**Review by jhaals — APPROVED** [review]
Looks good to me now! 👍 

Great if @benjdlambert or/and @acierto can do a final review as this is new APIs being added.

**Comment by drodil** [comment]
@jhaals I was thinking to wrap the offset/limit to pagination object still as it's how it's done in the catalog as well. Coming up with one more update to this!

File: plugins/scaffolder-backend/src/scaffolder/tasks/DatabaseTaskStore.ts [hunk 11]
```diff
@@ -183,20 +183,72 @@ export class DatabaseTaskStore implements TaskStore {
...
+        if (Array.isArray(filters.createdBy)) {
+          arr.push(...filters.createdBy);
+        } else {
+          arr.push(filters.createdBy);
+        }
+      }
+      queryBuilder.whereIn('created_by', [...new Set(arr)]);
+    }
+
+    if (status || filters?.status) {
```
Comment (acierto) [inline plugins/scaffolder-backend/src/scaffolder/tasks/DatabaseTaskStore.ts:214]: I see the pattern of dealing with the filter parameter for `status` or `createdBy`. It would be nice to create some util function which can transform it, and hide that processing noise to the reader of this function. 
Also it will help to create simple unit test for this transformation logic.

File: plugins/scaffolder-backend/src/service/helpers.ts [hunk 12]
```diff
@@ -107,3 +107,35 @@ export async function findTemplate(options: {
...
+  param: unknown,
+  paramName: string,
+): string[] | undefined {
+  if (param === undefined) {
+    return undefined;
+  }
+
+  const array = [param].flat();
+  if (array.some(p => typeof p !== 'string')) {
+    throw new InputError(`Invalid ${paramName}, not a string`);
```
Comment (acierto) [inline plugins/scaffolder-backend/src/service/helpers.ts:125]: It seems not fully correct error message. 
As the JsDoc saying that the array of strings is also a valid param.

File: plugins/scaffolder-backend/src/service/helpers.ts [hunk 13]
```diff
@@ -107,3 +107,35 @@ export async function findTemplate(options: {
...
+  }
+
+  return array as string[];
+}
+
+export function parseNumberParam(param: unknown, paramName: string) {
```
Comment (acierto) [inline plugins/scaffolder-backend/src/service/helpers.ts:131]: Though these functions are trivial, would be nice to have some unit tests for them to cover happy path and corner cases

File: plugins/scaffolder-backend/src/service/router.test.ts [hunk 14]
```diff
@@ -1198,11 +1221,22 @@ data: {"id":1,"taskId":"a-random-id","type":"completion","createdAt":"","body":{
...
+          pagination: {
+            limit: undefined,
+            offset: undefined,
+          },
+          filters: {
+            createdBy: undefined,
+            status: undefined,
+          },
+          order: undefined,
+          status: undefined,
```
Comment (acierto) [inline plugins/scaffolder-backend/src/service/router.test.ts:1239]: Wouldn't it be the same if to remove the explicit declaration of this params? (i believe all of them are optional)

File: plugins/scaffolder-backend/src/scaffolder/tasks/StorageTaskBroker.test.ts [hunk 15]
```diff
@@ -243,8 +244,10 @@ describe('StorageTaskBroker', () => {
 
     const task = await storage.getTask(taskId);
 
-    const promise = broker.list({ createdBy: 'user:default/foo' });
-    await expect(promise).resolves.toEqual({ tasks: [task] });
+    const promise = broker.list({
+      filters: { createdBy: ['user:default/foo'] },
+    });
+    await expect(promise).resolves.toEqual({ tasks: [task], totalTasks: 1 });
```
Comment (acierto) [inline plugins/scaffolder-backend/src/scaffolder/tasks/StorageTaskBroker.test.ts:250]: Would be nice to have a test for each new introduce parameter.
And some tests with a combination of them.

This test proves only that `createdBy` should work 😅

File: plugins/scaffolder-backend/src/scaffolder/tasks/DatabaseTaskStore.test.ts [hunk 16]
```diff
@@ -115,8 +115,34 @@ describe('DatabaseTaskStore', () => {
...
+      spec: {} as TaskSpec,
+      createdBy: 'me',
+    });
+
+    await store.createTask({
+      spec: {} as TaskSpec,
+      createdBy: 'him',
+    });
+
+    const { tasks, totalTasks } = await store.list({
```
Comment (acierto) [inline plugins/scaffolder-backend/src/scaffolder/tasks/DatabaseTaskStore.test.ts:141]: Same as in StorageTaskBroker it's nice to cover with tests all new introduced options

**Review by acierto — CHANGES_REQUESTED** [review]

File: plugins/scaffolder-backend/src/service/helpers.ts (hunk 12 above)
Comment (drodil) [reply plugins/scaffolder-backend/src/service/helpers.ts:125]: Sorry but what do you mean by this? 🤔 This function is copied from catalog utilities and works the same way there

File: plugins/scaffolder-backend/src/scaffolder/tasks/StorageTaskBroker.test.ts (hunk 15 above)
Comment (drodil) [reply plugins/scaffolder-backend/src/scaffolder/tasks/StorageTaskBroker.test.ts:250]: The other parameters are tested in other tests but sure I can add more explicit tests

File: plugins/scaffolder-backend/src/service/helpers.ts (hunk 12 above)
Comment (acierto) [reply plugins/scaffolder-backend/src/service/helpers.ts:125]: I mean that if `paramName` is an array of strings, it's still valid.

/**
 * Takes a single unknown parameter and makes sure that it's a single string or
 * an array of strings, and returns as an array.
 */

`throw new InputError(`Invalid ${paramName}, not a string or array of strings`);` would be more correct.

Could be also incorrect in catalog then 😅 

If I'm as a user will read the error and not check the code, I would have no clue that I can also provide the array of strings

**Comment by drodil** [comment]
@acierto cleaned up and added a punch of tests for the functionality!

**Review by acierto — APPROVED** [review]

File: plugins/scaffolder-backend/src/scaffolder/tasks/types.ts [hunk 17]
```diff
@@ -190,10 +190,34 @@ export interface TaskStore {
...
+  }): Promise<{ tasks: SerializedTask[]; totalTasks?: number }>;
+
+  /**
+   * @deprecated Make sure to pass `createdBy` and `status` in the `filters` parameter instead
+   */
   list?(options: {
     createdBy?: string;
     status?: TaskStatus;
-  }): Promise<{ tasks: SerializedTask[] }>;
+    filters?: {
```
Comment (benjdlambert) [inline plugins/scaffolder-backend/src/scaffolder/tasks/types.ts:211]: You don't have to add the additional fields in here, you can just trim this to the old options and the old API. This @deprecated one should be exactly the same as the old `list` method, and the new one above with the new options.

**Review by mounicarajput — APPROVED** [review]

**Comment by benjdlambert** [comment]
@drodil looks pretty good - just some failing tests I think :tada:

**Comment by Prateek-Wayne** [comment]
Waiting for this changes to be released.
Thanks @drodil
